Tiltfile

```diff
@@ -0,0 +1,123 @@
+# setup if you're repeatedly destroying and recreating your kind cluster, as it doesn't have to pull the images over
+# the network each time.
+def deploy_cert_manager():
+    registry = "quay.io/jetstack"
+    version = settings.get("cert_manager_version")
+    images = ["cert-manager-controller", "cert-manager-cainjector", "cert-manager-webhook"]
+
+    if settings.get("preload_images_for_kind"):
+        for image in images:
+            local("docker pull {}/{}:{}".format(registry, image, version))
+            local("kind load docker-image --name {} {}/{}:{}".format(settings.get("kind_cluster_name"), registry, image, version))
+
+    local("kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/{}/cert-manager.yaml".format(version))
+
+    # wait for the service to become available
+    local("kubectl wait --for=condition=Available --timeout=300s apiservice v1beta1.webhook.cert-manager.io")
+
+
+# deploy CAPI
+def deploy_capi():
+    version = settings.get("capi_version")
+    local("kubectl apply -f https://github.com/kubernetes-sigs/cluster-api/releases/download/{}/cluster-api-components.yaml".format(version))
+    # if settings.get("feature_gates"):
+    #     patch = {
+    #         "op": "add",
+    #         "path": "/spec/teamplate/spec/containers/1/args/1",
+    #         "value": settings.get("feature_gates"),
+    #     }
+    #     local("kubectl patch deployment capi-controller-manager -n capi-system  --type json -p='{}'".format(str(encode_json(patch))))
+
+
+# Users may define their own Tilt customizations in tilt.d. This directory is excluded from git and these files will
+# not be checked in to version control.
+def include_user_tilt_files():
+    user_tiltfiles = listdir("tilt.d")
+    for f in user_tiltfiles:
+        include(f)
+
+
+def append_arg_for_container_in_deployment(yaml_stream, name, namespace, contains_image_name, arg):
+    for item in yaml_stream:
+        if item["kind"] == "Deployment" and item.get("metadata").get("name") == name and item.get("metadata").get("namespace") == namespace:
+            containers = item.get("spec").get("template").get("spec").get("containers")
+            for container in containers:
+                if contains_image_name in container.get("image"):
+                    container.get("args").append(arg)
+
+
+def fixup_yaml_empty_arrays(yaml_str):
+    yaml_str = yaml_str.replace("conditions: null", "conditions: []")
+    return yaml_str.replace("storedVersions: null", "storedVersions: []")
+
+
+# Enable core cluster-api plus everything listed in 'enable_providers' in tilt-settings.json
+def capz():
+    # Apply the kustomized yaml for this provider
+    yaml = str(kustomize("./config"))
+    substitutions = settings.get("kustomize_substitutions", {})
+
+    keys = ["AZURE_SUBSCRIPTION_ID_B64", "AZURE_TENANT_ID_B64", "AZURE_CLIENT_SECRET_B64", "AZURE_CLIENT_ID_B64"]
```

Comment:
can we check for missing env vars earlier? ie. before deploying the kind cluster?

Reply:
also (if it's not asking for too much) maybe we can automate exporting the base 64 version of the variables

Reply:
hmm actually just realized this has to be part of tilt-settings.json, not exported as env vars

Reply:
We could pick them up from env probably. I just followed the same thing that was already there.

Reply:
I'll move validation earlier in the flow.